modules/vat/model/mvaavstemming.class.php

```diff
@@ -41,29 +41,34 @@ function mva_avstemming($args)
-            //print "XXXX: $this->_query<br>";
-
-            $this->_row     = $this->_dbh[$this->_dsn]->get_row(array('query' => $this->_query));
-            $this->_outAccountPlanID[$vatpercent] = $this->_row->AccountPlanID;
-
-            #$this->_query   = "select AccountPlanID from vat where Percent='$vatpercent' order by VatID desc";
-            $this->_query    = "select c.AccountPlanID from voucher as c, voucher as m where m.VoucherPeriod like '" . $this->year . "%' and m.AutomaticVatVoucherID=c.VoucherID and m.Vat='" . $vatpercent . "' and m.VatID >= 40 and m.VatID <= 60 and m.Active=1 and c.Active=1 group by c.AccountPlanID"; # and AmountOut > 0
-
-            $this->_row     = $this->_dbh[$this->_dsn]->get_row(array('query' => $this->_query));
-            $this->_inAccountPlanID[$vatpercent] = $this->_row->AccountPlanID;
+        $vat_registers_query = "SELECT * FROM vat WHERE Percent != 0 AND Percent IS NOT NULL ORDER BY VatID";
+        $vat_accounts_used_in_vouchers_query = "SELECT DISTINCT(AccountPlanID) AS AccountPlanID FROM voucher WHERE VoucherPeriod LIKE '{$this->year}%' AND AutomaticVatVoucherID != 0 AND Active = 1 AND DisableAutoVat = 1";
+        $vat_registers = $this->_dbh[$this->_dsn]->get_hashhash(array('query' => $vat_registers_query, 'key' => 'ID'));
+        $vat_accounts_used_in_vouchers = $this->_dbh[$this->_dsn]->get_hashhash(array('query' => $vat_accounts_used_in_vouchers_query, 'key' => 'AccountPlanID'));
+        foreach ($vat_registers as $vat_register) {
+          $report_time = strtotime($this->year.'-12-31'); // Compare with the last day of the selected report's year
+          $valid_to_time = strtotime($vat_register['ValidTo']);
+          $valid_from_time = strtotime($vat_register['ValidFrom']);
+          if ($report_time > $valid_to_time || $report_time < $valid_from_time) {
+            continue;
+          }
+          // needs to be a string with float foramted to two decimal places in
+          // order for the correct fields to be picked up in the report
+          $vat_percent = number_format((float)$vat_register['Percent'], 2, '.', '');
+          if (($vat_register['Active'] || !empty($vat_accounts_used_in_vouchers[$vat_register['AccountPlanID']])) && $vat_register['VatID'] <= 39) {
+            $this->_outAccountPlanID[$vat_percent] = $vat_register['AccountPlanID'];
+            $this->vathash[$vat_percent] = $vat_register;
+          }
+          if (($vat_register['Active'] || !empty($vat_accounts_used_in_vouchers[$vat_register['AccountPlanID']])) && $vat_register['VatID'] > 39) {
+            $this->_inAccountPlanID[$vat_percent] = $vat_register['AccountPlanID'];
+            $this->vathash[$vat_percent] = $vat_register;
```

Comment:
The value of this hash is not used afterwards. Use Vat percent same as it was in the query used before?